L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,369 @@
+1. support for certificate revocation via Certificate Revocation Lists (CRLs)
+
+## Background
+
+The current TLS implementation in gRPC C core has some restrictions:
+1. lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down
+1. inflexibility in the peer verification. In a mutual TLS scenario, servers would always perform certificate verification, and clients would always perform certificate verification as well as default hostname check. No customized checks can be introduced.
+1. inability to choose the maximum/minimum TLS version supported in OpenSSL/BoringSSL for use
+
+There have always been some ad hoc requests on GitHub Issue Page for supporting these features. The demands for a more flexible and configurable API have increased with the emerging needs in the cloud platform.
+
+One common requirement is to skip the hostname verification check if a client has a different way of validating the server's identity in different platforms. 
+For example, it's becoming more and more popular to use SPIFFE ID as the identity format, but a valid certificate using SPIFFE ID doesn't necessarily contain the dnsName in the SAN field, which might cause hostname verification to fail.
+
+Another requirement is to reload the root and identity credentials without shutting down the server, because certificates will eventually expire and the ability to rotate the certificates is essential to a cloud application.
+
+Besides requests from Open Source Community, with the growth of GCP, we've seen more and more internal use cases of these features.   
+
+Hence we propose an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones doesn't fit their needs
+1. The credentials reloading should be efficient. We shall reload only when needed rather than reloading per TLS connection
+1. C core API needs to be simple enough, so that most of existing SSL credentials in wrapped languages can migrate to call this new API
+
+### Related Proposals:
+
+1. [L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98) - this was subsumed into this proposal.
+1. [A69: Crl Providers](https://github.com/grpc/proposal/pull/382)
+
+## Proposal
+
+### TLS credentials and TLS Options
+This part of the proposal introduces `TlsCredentialsOptions` and TLS credentials. Users use `TlsCredentialsOptions` to configure the specific security properties they want, and build the client/server credential by passing in `TlsCredentialsOptions`.
+This section only covers the configuration not related to credential reloading and custom verification. Configurations related to those two topics are explained in the latter sections.
```

Comment:
Rather than having this sentence, can we structure this as e.g. "Basic structure" or "Basic options", and then cover credential reloading and custom verification in subsequent sections?

Reply:
Removed this sentence, I think it just makes more sense without it